### PR TITLE
GHA: build master w/ deps' masters, not releases (again)

### LINKED ISSUE #62: GHA: build master w/ deps' masters, not releases (again)
## Blocked by

* #61

### LINKED ISSUE #78: icinga/icingaweb2:master css build failed
In the current master on docker hub the build for icinga.min.css seems to have failed and instead returns this error:

```less
.appearance is undefined in anonymous-file-0.less
#0 /usr/share/icingaweb2/library/vendor/lessphp/lib/Less/Tree/Ruleset.php(143): Less_Tree_Mixin_Call->compile()
#1 /usr/share/icingaweb2/library/vendor/lessphp/lib/Less/Tree/Ruleset.php(86): Less_Tree_Ruleset->EvalMixinCalls()
#2 /usr/share/icingaweb2/library/vendor/lessphp/lib/Less/Tree/Ruleset.php(91): Less_Tree_Ruleset->compile()
#3 /usr/share/icingaweb2/library/vendor/lessphp/lib/Less/Tree/Ruleset.php(91): Less_Tree_Ruleset->compile()
#4 /usr/share/icingaweb2/library/vendor/lessphp/lib/Less/Parser.php(185): Less_Tree_Ruleset->compile()
#5 /usr/share/icingaweb2/library/vendor/lessphp/lessc.inc.php(130): Less_Parser->getCss()
#6 /usr/share/icingaweb2/library/Icinga/Web/LessCompiler.php(230): lessc->compile()
#7 /usr/share/icingaweb2/library/Icinga/Web/StyleSheet.php(207): Icinga\Web\LessCompiler->render()
#8 /usr/share/icingaweb2/library/Icinga/Web/StyleSheet.php(247): Icinga\Web\StyleSheet->render()
#9 /usr/share/icingaweb2/library/Icinga/Application/webrouter.php(64): Icinga\Web\StyleSheet::send()
#10 /usr/share/icingaweb2/public/index.php(4): require_once('/usr/share/icin...')
#11 {main}


.. em;
       width: 1.5em;
     }
   }
   
   .service-grid-link {
     .bg-stateful();
     .rounded-corners();
   
     display: inline-block;
     height: 1.5em;
     vertical-align: middle;
     width: 1.5em;
   }
   }/* Icinga Web 2 | (c) 2021 Icinga GmbH | GPLv2+ */
   
   @enable-color-preference: 999999px;
   @prefer-light-color-scheme: 999999px;
   
..
```

Do you need other input to track this down?

### PR DESCRIPTION
**Pull request by Al2Klimov**
fixes #62
closes #78

### CODE REVIEW

**Comment by Al2Klimov** [comment]
Seems to work: https://github.com/Al2Klimov/icingaweb2/runs/4819983828?check_suite_focus=true#step:3:545

**Review by nilmerg — CHANGES_REQUESTED** [review]
This only works as long as the default branch of each module matches the one of icingaweb2. (master <> main)

**Comment by julianbrost** [comment]
There's already some special-casing for `BRANCH=master` in `get-mods.sh`. I don't think you won't get this right for arbitrary branches unless you explicitly specify the dependency versions for each separately.

**Review by nilmerg — CHANGES_REQUESTED** [review]
Has something changed? No? My point stands. There's no special dependency management needed, just a fallback procedure in case `master` is not valid.

**Comment by Al2Klimov** [comment]
I thought GitHub... branch mirroring? will do this for us.

**Comment by nilmerg** [comment]
[In which way is Github involved here?](https://github.com/Icinga/docker-icingaweb2/blob/master/get-mods.sh#L39)

**Comment by julianbrost** [comment]
@nilmerg So you're asking for another case in `get-mods.sh` where if we're looking for `master`, but `master` does not exist but `main` does, use `main` instead? Is that relevant for any existing repos or are you planning to create any such repo soon?

**Comment by Al2Klimov** [comment]
C'mon, pause the discussion for now. **Other construction area.**

If some people in our department want to reduce the racism here from 0% to.. 0% by s/master/main/ – fine. They are free to go for it. **They.** I've requested **your** review to get **this construction area** done.

Do **you** have any complaints or not?

**Comment by julianbrost** [comment]
C'mon, I just want to understand the review before I'd outvote it.

**Comment by nilmerg** [comment]
I'm really baffled. This isn't just for the case *that if someone renames the default branch*. (Which we have already talked about to do for Icinga DB Web) `main` is the *default* default branch of Github for new repositories. I point this case now out, and [it's a simple to solve problem](https://github.com/Icinga/L10n/blob/main/bin/update#L42-L46), why should we wait and have to solve it sometimes later? This is about that everywhere the latest nightly stage is used, be it `master` or `main`.

*I* won't give a positive review for this until this is solved. And it's my last comment regarding this.

**Comment by julianbrost** [comment]
I'm asking because I'm confused about the branch naming. For the newest two repos I'm currently aware of (`icinga-php-{library,thirdparty}`), the branch is supposed to be `snapshot/nightly`, suggesting some fully custom branch naming. Is this some we'll see more often now? Will this stay a special case and new things now get `main` instead since you're asking for that specifically?

> And it's my last comment regarding this.

Well, if I don't understand the branch naming, I can't give a review beyond "Does it look like it might work now? Maybe, I guess?".

**Comment by nilmerg** [comment]
There are special cases like `snapshot/nightly`, if there will be more like this in the future is unknown. But these are really **special** cases since there is still a default branch, it just doesn't contain the latest stage. The *not-so-special* case `main` **is known** and **will** be used in the future for all new repositories, that's what we've settled on if I'm not mistaken. And I guarantee Icinga DB Web will change it, then this breaks and here we are again.

I really don't understand why it's so difficult to plan for this now and why we even discuss this that far. If you don't want to do it, fine, just don't blame me if some repository get's a different default branch and the docker images are broken again.

**Comment by julianbrost** [comment]
We also briefly discussed this in our core meeting and concluded to just allow for both branch names, even if not strictly necessary right now to avoid having this break yet again in the near future.

File: get-mods.sh [hunk 1]
```diff
@@ -25,7 +25,7 @@ get_special () {
 				if [ "$BRANCH" = master ] && [[ "$2" == icinga-php/* ]]; then
 					REF=snapshot/nightly
 				elif [ -n "$BRANCH" ] && git -C dockerweb2-temp show -s --oneline "$BRANCH"; then
-					REF="$BRANCH"
+					REF=HEAD
```
Comment (julianbrost) [inline get-mods.sh]: Doesn't the following describe the change? If we have a branch name and that branch exists, use HEAD instead of that branch name.

That doesn't sound right to me.

File: action.bash [hunk 2]
```diff
@@ -64,7 +64,12 @@ case "$GITHUB_EVENT_NAME" in
 		else
 			TAG="${BASH_REMATCH[2]}"
 		fi
-		mkimg
+
+		if [ "${BASH_REMATCH[1]}" = heads ]; then
+			BRANCH="${BASH_REMATCH[2]}"
+		fi
```
Comment (julianbrost) [inline action.bash:70]: Can you please explain the reason for this change?

The whole case block is somewhat confusing and I'd find it much more helpful to document this block rather than what the regex does in detail.

Comment (Al2Klimov) [reply action.bash:70]: Aren't both the condition (is $REF a branch?) and the action (use that branch) obvious (w/o even knowing the array positions in advance)?

File: get-mods.sh [hunk 3]
```diff
@@ -24,8 +24,8 @@ get_special () {
 			*)
 				if [ "$BRANCH" = master ] && [[ "$2" == icinga-php/* ]]; then
 					REF=snapshot/nightly
-				elif [ -n "$BRANCH" ] && git -C dockerweb2-temp show -s --oneline "$BRANCH"; then
-					REF="$BRANCH"
+				elif [ -n "$BRANCH" ]; then
+					REF=HEAD
```
Comment (julianbrost) [inline get-mods.sh:28]: Now it loses the feature to use matching branches from all repositories. Is this intentional?

**Review by julianbrost — COMMENTED** [review]
Can you please explain in your own words (i.e. not just paraphrasing the code), what should happen (Which branches are used from which repos? Is the image pushed to Docker Hub? If yes, using which name/tags?) when the action is run on tags or different branches (`master`, `support/*`, `feature/*`).

File: get-mods.sh (hunk 3 above)
Comment (Al2Klimov) [reply get-mods.sh:28]: Yes.

**Comment by Al2Klimov** [comment]
OK, actually it's damn simple:

* master -> HEAD (or nightly)
* tag -> tag

**Comment by julianbrost** [comment]
That doesn't fully answer my question. What happens for other branches? It clones from multiple repos, that have different branches and tags, what is used for each? Also you said nothing about tagging and uploading to Docker Hub.

**Comment by Al2Klimov** [comment]
Which other branches?

https://github.com/Icinga/icingaweb2/blob/42963d1c1e41f424ac14e60011563d569c7a6b65/.github/workflows/docker.yml#L3-L10

**Comment by julianbrost** [comment]
The [documentation for `workflow_dispatch`](https://docs.github.com/en/actions/managing-workflow-runs/manually-running-a-workflow#running-a-workflow) looks like it's possible to trigger it for any branch.

The point is that to me, it's unclear what this action is supposed to support in the end. You propose to [remove some functionality](https://github.com/Icinga/docker-icingaweb2/pull/73#discussion_r798756558) of which I don't if someone relies on as I don't work on our web components, that's why I was asking for some kind of specification.

**Comment by Al2Klimov** [comment]
For other branches it will choose HEAD.

**Comment by Al2Klimov** [comment]
@nilmerg PING

**Comment by Al2Klimov** [comment]
See also my last OP edit.

**Comment by lippserd** [comment]
If @julianbrost is fine with this, please just merge it.

**Comment by julianbrost** [comment]
> The point is that to me, it's unclear what this action is supposed to support in the end. You propose to [remove some functionality](https://github.com/Icinga/docker-icingaweb2/pull/73#discussion_r798756558) of which I don't if someone relies on as I don't work on our web components, that's why I was asking for some kind of specification.

I mean that's still not clear to me. If the only requirement is that this repo can build the `master` and `x.y.z` release images, then it might work (I looked at this some time ago and I hope I would have commented if that core functionality was broken), but if these are all the requirements, I think the code is way to complex to achieve this.

**Comment by Al2Klimov** [comment]
$OTHER_IW2_BRANCH also works, but that chooses $MODULE->HEAD.

**Review by julianbrost — COMMENTED** [review]
However, for branches other than `master`, it will also choose `HEAD` for `icinga-php/*`, which will - as far as I understand - not work at all because otherwise we wouldn't need to special-case that to `snapshot/nightly` in the first place.

Will anyone ever use this for other branches from GitHub? I don't know.
Will anyone ever care if this works? I don't know.
Should I know? Only if someone tells me, I don't actively work on Icinga Web 2.

So in summary, I don't think I can give a meaningful review unless someone finally explains the expected feature set of this repository to me.

I have just looked at the changes again. I think it will work for `master` and release tags. I consider everything else undefined behavior for now. If that's good enough, go ahead and merge it, but don't blame me if something breaks that I had no idea anyone expected to work.

**Comment by nilmerg** [comment]
> Will anyone ever use this for other branches from GitHub? I don't know.

Since we don't use the image in our internal dev-environment anyway, no-one. If a user or customer comes around who want's to test a specific change (PR) in a docker environment, it might come in handy. But this never occurred. But maybe this changes once the image works fine?

I'm fine with whatever solution, as long as the github action is adjusted accordingly in the icingaweb2 repository as well.

**Comment by julianbrost** [comment]
What do you think of the following suggestions?
1. Change `get-mods.sh` so that its parameter is called/described as something like build mode or dependency versions. It may take two values: `release` (all modules used in the image use their latest release) and `snapshot` (all modules use the snapshot/nightly, master, main, HEAD, or whatever it's called in each repo).
2. Change `action.bash` so that it sets this build mode to `snapshot` for `refs/heads/*` and to `release` for `refs/tags/*`.
3. (Maybe, I'm unsure what the expectations are) Restrict pushes to Docker Hub to `refs/heads/master` and `refs/tags/v*`.
4. (Optional, but might make sense to just implement it now) Expose this build mode as an action input so that the workflow file can explicitly set it. Use something like `auto` as default which uses the rules described in (2). For example, this would then allow building Docker Images for `support/2.9` with release dependency versions.
5. (Optional, probably only makes sense to implement if actually needed at some point) Expose an additional action input that can be used to override the version for specific modules.

**Comment by nilmerg** [comment]
1. Okay, but the ref itself is still passed?
2. Okay
3. Conflicts with 4.
4. :+1: 
5. Must be possible if 4. is done

**Comment by julianbrost** [comment]
1. I don't see a reason why this would be needed. But as that value comes from an environment variable anyways, this can just be accessed directly (but I wouldn't access it in `get-mods.sh`)
2. (intentionally left blank as GitHub doesn't seem to allow skipping numbers in lists)
3. I don't see how this conflicts. But I think this may depend on how this action is viewed. Is it only intended for building images with the purpose of uploading them or is there also value in just running the build for testing purposes. This probably differs between icinga2 and icingaweb2. So this is all about the question if anyone would be surprised finding out later that this action automatically uploaded this image to Docker Hub when they only wanted to test something on a branch.
4. (intentionally left blank as GitHub doesn't seem to allow skipping numbers in lists)
5. You mean should be implemented if (4) is implemented? In my understanding, that would be a mechanism for something like the override for the icingadb-web 1.0.0rc1 that was in the repo before, so nothing that's needed right away.

**Comment by nilmerg** [comment]
1. I didn't take a look at the code, but from my understanding, if it only gets the mode, which ref is built in case of `release`? (i.e. which release)
2. 
3. support/2.9 is not a tag, hence if pushes are restricted to just refs/heads/master there is no new image pushed for 2.9. We intentionally decided not to build docker images in PRs for web components as this is useless. (cross-compatibility and so on)
4.
5. If a build is ran for e.g. support/2.9, a module that requires Icinga Web 2.11 must not be installed

**Comment by julianbrost** [comment]
This build mode would only affect the dependencies/modules downloaded by `get-mods.sh`, not the version of icingaweb2 (there it always takes whatever version the action in run on). And for these, it always uses the latest release.

But this brings up another issue: this repo currently does not allow using a different set of modules for different releases. So if the situation in (5) happens, you couldn't release 2.11.0 and 2.9.x and only have that new module in the 2.11.x image without special-casing this in `get-mods.sh`.

**Comment by nilmerg** [comment]
Yeah. Module packages are on the way anyway. Just let's drop this here and use packages instead.

**Comment by Al2Klimov** [comment]
Don't they depend on the IW2 package?

**Comment by nilmerg** [comment]
...

release = use packages (including Icinga Web 2), should include builds for support branches
snapshot = use latest everywhere (master, main, snapshot/nightly, ...)

profit.

**Comment by lippserd** [comment]
I don't think it's a solution to completely change the way how docker containers work and "just" install a package.

**Comment by nilmerg** [comment]
Then don't ask me any further how to solve what. I've made my requirements clear.

**Comment by lippserd** [comment]
While the discussion isn't over here, the PR fixes the issue mentioned, so I'll merge it.

**Comment by dwt** [comment]
> While the discussion isn't over here, the PR fixes the issue mentioned, so I'll merge it.

Just curious - are you waiting for #79 to merge, or is there another reason for the holdup for now?
